wrapper.py

```diff
@@ -37,15 +37,33 @@
 # 3. numpy type casting rules (to be cleaned up in numpy: follow old or new)
 
 
+
+def asarray_replacer_1(func):
+    """Asarray the *first* arg, so that the wrapped `func` operates on a
+       torch.Tensor. Then asarray the result.
+    """
+    def wrapped(x, *args, **kwds):
+        x_tensor = asarray(x).get()
+        return asarray(func(x_tensor, *args, **kwds))
+    return wrapped
+
+
+
+
 NoValue = None
 
 ###### array creation routines
 
 def asarray(a, dtype=None, order=None, *, like=None):
-    _util.subok_no_ok(like)
+    _util.subok_not_ok(like)
     if order is not None:
         raise NotImplementedError
-    return torch.asarray(a, dtype=dtype)
+
+    # This and array(...) are the only places which talk to ndarray directly.
+    # The rest goes through asarray (preferred) or array.
+    out = ndarray()
+    out._tensor = torch.asarray(a, dtype=dtype)
```

Comment:
@lezcano could you enlighten me on the difference between torch.asarray and torch.as_tensor?

Reply:
They are equivalent. cf.
https://github.com/pytorch/pytorch/blob/e54c6c2870d6f1e318cdd86fa94c37ad22bff22a/test/test_tensor_creation_ops.py#L3908-L3911

Reply:
Thanks! Here's one puzzler though:

```
In [43]: x = np.float64(0.5)

In [44]: torch.asarray(x)
<ipython-input-44-62b3e1ccf04c>:1: UserWarning: The given buffer is not writable, and PyTorch does not support non-writable tensors. This means you can write to the underlying (supposedly non-writable) buffer using the tensor. You may want to copy the buffer to protect its data or make it writable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at /home/conda/feedstock_root/build_artifacts/pytorch-recipe_1670076225403/work/torch/csrc/utils/tensor_new.cpp:1563.)
  torch.asarray(x)
Out[44]: tensor([0.0000, 1.7500])

In [45]: torch.as_tensor(x)
Out[45]: tensor(0.5000, dtype=torch.float64)
```

Is is supposed to handle numpy scalars?

Reply:
No idea about that edge case, but if  `as_tensor` works, I'd simply use `torch.as_tensor`.
cc @ysiraichi who wrote the test I linked above.

Reply:
Not an edge case, but a bug! It should work fine if you specify the dtype, though.

```
>>> torch.asarray(x, dtype=torch.float64)
tensor([0.5000], dtype=torch.float64)
```

Reply:
Correction: it's not a bug... I will look into this.

```
In [6]: torch.asarray(np.array([1, 2, 3], dtype=np.float64))
Out[6]: tensor([1, 2, 3])
```

Reply:
Long story short: `torch.asarray` works with Numpy scalars, but it doesn't infer the type. Thus, if you don't specify the dtype, it will assume the default dtype.

**Explanation:** since Numpy scalars are not Numpy arrays, they end up being treated as Python objects that implement the protocol buffer ([as documented](https://pytorch.org/docs/stable/generated/torch.asarray.html)). i.e. it will rely on the given `dtype` argument (if given) and interpret the data as such.

**Discussion:** I guess the whole point of `asarray` is to try not to copy. I don't think it makes much sense using it for scalars... That said, it's an non-intuitive result that has an easy fix.

@lezcano @rgommers 
What do you think?

Reply:
I reckon that extending `torch.asarray` to also handle numpy scalars as expected wouldn't hurt, but let's see what Ralf has to say.